provider/lxd/environ.go

```diff
@@ -470,31 +471,47 @@ func (env *environ) DetectSeries() (string, error) {
 
 // DetectHardware returns the hardware characteristics for the controller for
 // this environment. This method is part of the environs.HardwareCharacteristicsDetector
-// interface. On an LXD cloud, it must first check if it is a localhost cloud,
-// in that case, we only fill the Arch.
+// interface. On an LXD cloud, it must first check if it is a local cloud, in
+// that case, we only fill the Arch constraint.
+// This method should always return nil errors, because it is run during
+// bootstrap and its only purpose is to update the Arch constraint so in case
+// of an error the default arch is used and the bootstrap can continue.
 func (env *environ) DetectHardware() (*instance.HardwareCharacteristics, error) {
-	// The returned error is willingly ignored, because this should not
-	// break bootstrapping (we detect hardware before bootstrapping),
-	// instead, bootstrapping should fallback to default hardware arch.
-	endpointURL, _ := url.Parse(env.cloud.Endpoint)
+	// In order to determine if this is a local lxd cloud, we need to
+	// extract its endpoint IP.
+	// The endpoint can be formatted either as https://host:port,
+	// https://host, host:port, host, https://IP:port, https://IP,
+	// IP:port, or only IP, so we try to parse it as a URL but first
+	// ensuring it contains the correct scheme.
+	endpointURL, err := url.Parse(lxd.EnsureHTTPS(env.cloud.Endpoint))
```

Comment:
I'm happy enough for this to land.

I will note that inside of *our* heads, we should be clear that Endpoint is not always an lxd endpoint. It is also used for every other provider (ec2 endpoint, azure endpoint, etc)

It does happen that *DetectHardware* only cares about lxd endpoints (and we do document that in the DetectHardware comment).

I do wonder if we'll eventually regret this special casing of one particular thing...